File: typescript/packages/core/src/auth/oauth2-config.ts
/**
 * OAuth2 Configuration Types - PROVIDER-AGNOSTIC + YAML-DRIVEN
 *
 * Matimo supports OAuth2 for any provider via YAML configuration.
 * Provider definitions are loaded from tools/[provider]/definition.yaml files.
 *
 * Matimo's Responsibility:
 * - Help obtain tokens via standard OAuth2 flow
 * - Support automatic token refresh
 * - Work with Google, GitHub, Slack, or any OAuth2 provider
 * - Load provider configs from YAML (no hardcoded endpoints)
 *
 * User's Responsibility:
 * - Store tokens (DB, file, cache, wherever you want)
 * - Retrieve and pass tokens to Matimo tools
 *
 * Adding a New Provider (e.g., Microsoft Teams):
 * 1. Create: tools/microsoft/definition.yaml
 * 2. Add provider section with OAuth endpoints
 * 3. Done! OAuth2Handler automatically recognizes it.
 *
 * Configuration Priority (highest to lowest):
 * 1. config.endpoints - Runtime override (user provides directly)
 * 2. OAUTH_PROVIDER_AUTH_URL env var - Deployment override
 * 3. YAML definition (tools/[provider]/definition.yaml) - Default from config
 */

/**
 * Supported OAuth2 providers
 *
 * Type is string (not enum) because providers are discovered from YAML.
 * Tools define which provider they use via authentication.provider field.
 */
export type OAuth2Provider = string;

/**
 * OAuth2 token - Returned by OAuth2Handler
 *
 * This is what you get after successful authorization.
 * Store this in your DB/file/cache (Matimo doesn't store it).
 * Pass accessToken to any Matimo tool that needs OAuth.
 */
export interface OAuth2Token {
  accessToken: string; // Use this: await matimo.execute(tool, { OAUTH_TOKEN: accessToken })
  refreshToken?: string; // Optional: for refresh flow
  expiresAt: number; // Unix timestamp in milliseconds
  scopes: string[]; // What this token can access
  provider: OAuth2Provider; // Which provider issued it
  userId: string; // Associated user ID
}

/**
 * OAuth2 configuration - PROVIDER-AGNOSTIC + YAML-DRIVEN
 *
 * Same interface works for Google, GitHub, Slack, or any OAuth2 provider.
 * Provider endpoints come from YAML definitions (tools/[provider]/definition.yaml).
 * Users can override endpoints without touching code.
 *
 * Configuration Priority:
 * 1. endpoints property (runtime override - user provides directly)
 * 2. OAUTH_PROVIDER_AUTH_URL environment variable (deployment config)
 * 3. YAML definition from tools/[provider]/definition.yaml (default configuration)
 */
export interface OAuth2Config {
  provider: OAuth2Provider; // Provider name (e.g., 'google', 'github')
  clientId: string; // From OAuth2 provider console
  clientSecret: string; // From OAuth2 provider console
  redirectUri: string; // Callback URL after authorization
  endpoints?: OAuth2Endpoints; // Optional: override YAML or env vars (highest priority)
}

/**
 * Authorization request options
 */
export interface AuthorizationOptions {
  scopes: string[];
  userId: string;
  state?: string;
}

/**
 * Token exchange response from OAuth2 provider
 */
export interface TokenResponse {
  access_token: string;
  refresh_token?: string;
  expires_in: number; // Seconds
  token_type: string;
  scope?: string;
}

/**
 * OAuth2 endpoints for a provider
 */
export interface OAuth2Endpoints {
  authorizationUrl: string;
  tokenUrl: string;
  revokeUrl?: string;
}

/**
 * NOTE: Provider endpoints are loaded from YAML files, not hardcoded here.
 * See: tools/[provider]/definition.yaml for provider configurations
 *
 * This enables:
 * - Infinite provider support
 * - Configuration-driven design
 * - Easy provider additions without code changes
 * - Centralized endpoint management
 */

File: python/packages/core/src/matimo/auth/oauth2_config.py
"""
OAuth2 Configuration Types — PROVIDER-AGNOSTIC + YAML-DRIVEN.

Mirrors: packages/core/src/auth/oauth2-config.ts

Matimo supports OAuth2 for any provider via YAML configuration.
Provider definitions are loaded from tools/[provider]/definition.yaml.

Configuration Priority (highest → lowest):
  1. ``config.endpoints`` — runtime override (user provides directly)
  2. ``OAUTH_PROVIDER_AUTH_URL`` env var — deployment override
  3. YAML definition (tools/[provider]/definition.yaml) — package default
"""
from __future__ import annotations

from pydantic import BaseModel, Field

# Re-export the endpoint model from core (single source of truth)
from matimo.core.models import OAuth2Endpoints

__all__ = [
    "OAuth2Endpoints",
    "OAuth2Config",
    "AuthorizationOptions",
    "TokenResponse",
    "OAuth2Token",
]


class OAuth2Config(BaseModel):
    """
    Provider-agnostic OAuth2 configuration.

    Same interface works for Google, GitHub, Slack, etc.
    Provider endpoints come from YAML definitions.
    """

    provider: str
    client_id: str
    client_secret: str
    redirect_uri: str
    endpoints: OAuth2Endpoints | None = None


class AuthorizationOptions(BaseModel):
    """Options for generating an authorization URL."""

    scopes: list[str]
    user_id: str
    state: str | None = None


class TokenResponse(BaseModel):
    """Raw token response from the OAuth2 provider."""

    access_token: str
    refresh_token: str | None = None
    expires_in: int  # seconds
    token_type: str = "Bearer"  # noqa: S105
    scope: str | None = None


class OAuth2Token(BaseModel):
    """
    Structured token returned by OAuth2Handler.

    Store this in your DB/file/cache — Matimo does NOT store tokens.
    Pass ``access_token`` to any Matimo tool that needs OAuth.
    """

    access_token: str
    refresh_token: str | None = None
    expires_at: int  # Unix timestamp in *milliseconds*
    scopes: list[str] = Field(default_factory=list)
    provider: str
    user_id: str
